core/unix/signal.c

```diff
@@ -6693,6 +6700,141 @@ execute_native_handler(dcontext_t *dcontext, int sig, sigframe_rt_t *our_frame)
+// which is assumed to be at xsp. If it succeeds, it does not return. Otherwise
+// it returns a bool: false if DR should continue trying to deliver the signal.
+static bool
+execute_native_handler_using_cur_frame(dcontext_t *dcontext, int sig, byte *xsp)
+{
+    // For a signal that arrived when the thread is native and the process is in
+    // the middle of detaching, we should try to reuse the current DR signal
+    // frame to deliver it if possible. This is to avoid using up too much app
+    // stack, which has been seen to cause a stack overflow in the app (i#6814).
+    // If we cannot optimize, we fall back to the regular
+    // execute_native_handler() on returning.
+    //
+    // Summary of relevant events when DR starts detaching:
+    // D: Detacher thread sends suspend signal to all other threads
+    // O: each of the Other threads receive the suspend signal. Note that DR's
+    //    signal handling configuration blocks all signals except SIGSEGV and
+    //    the suspend signal. So at this point Other threads cannot receive
+    //    app signals. Other threads signal to the Detacher thread that they're
+    //    suspended and awaiting wakeup call to continue detaching (sig_detach).
+    // D: Detacher thread sets doing_detach to true. A side effect of this is
+    //    that get_thread_private_dcontext() will return NULL when any Other
+    //    thread tries to get its private dcontext.
+    // D: Detacher thread wakes up each Other thread, telling it to detach.
+    // O: Each Other thread wakes up, does sig_detach which reinstates the app
+    //    signal stack (if available) and the app's blocked signal set, lets the
+    //    Detacher thread know that its done detaching and resumes native
+    //    execution. Note that at this point, DR's sigact config is still
+    //    installed (including main_signal_handler). When the Other thread
+    //    return from the suspend signal, signals get unblocked automatically
+    //    and may now be delivered.
+    // <execute_native_handler_using_cur_frame may help here>
+    // D: When all threads are done detaching, the Detacher thread cleans up
+    //    each thread's drcontext for real, and also reinstates the app's
+    //    sigact in signal_thread_exit (which removes main_signal_handler).
+    //
+    // Therefore, if we arrive here and find doing_detach to be true and
+    // dcontext to be NULL, we can be certain that the current thread has
+    // completed sig_detach. This means the interrupted context stack (sc->XSP)
+    // must be the app's own regular stack.
+    //
+    // Also, if the thread has a signal stack at all for any signal, we will
+    // be on it (because DR sets SA_ONSTACK for all signals, and that config is
+    // still in effect), regardless of whether the signal stack was registered
+    // for sig. If the signal stack was actually not registered for sig, then we
+    // return true (indicating that DR should continue trying to deliver the
```

Comment:
This contradicts above where it says false means to continue trying.

Reply:
Oops. Fixed it.